kernel/src/dynamic_process_loading.rs

```diff
@@ -0,0 +1,958 @@
+/// The client capsule should implement this trait to handle the callback logic
+pub trait DynamicProcessLoadingClient {
+    fn app_data_write_done(&self, buffer: &'static mut [u8], length: usize);
+}
+
+pub struct DynamicProcessLoader<'a, C: 'static + Chip> {
+    kernel: &'static Kernel,
+    chip: &'static C,
+    fault_policy: &'static dyn ProcessFaultPolicy,
+    procs: MapCell<&'static mut [Option<&'static dyn process::Process>]>,
+    flash: Cell<&'static [u8]>,
+    app_memory: OptionalCell<&'static mut [u8]>,
+    flash_start: Cell<usize>,
+    flash_end: Cell<usize>,
+    new_process_flash: OptionalCell<&'static [u8]>,
+    driver: &'a dyn NonvolatileStorage<'a>,
+    buffer: TakeCell<'static, [u8]>,
+    new_app_start_addr: Cell<usize>,
+    new_app_length: Cell<usize>,
+    client: OptionalCell<&'static dyn DynamicProcessLoadingClient>,
+    current_user: OptionalCell<NonvolatileUser>,
+}
+
+impl<'a, C: 'static + Chip> DynamicProcessLoader<'a, C> {
+    pub fn new(
+        processes: &'static mut [Option<&'static dyn process::Process>],
+        kernel: &'static Kernel,
+        chip: &'static C,
+        flash: &'static [u8],
+        fault_policy: &'static dyn ProcessFaultPolicy,
+        driver: &'a dyn NonvolatileStorage<'a>,
+        buffer: &'static mut [u8],
+    ) -> Self {
+        Self {
+            procs: MapCell::new(processes),
+            kernel,
+            chip,
+            flash: Cell::new(flash),
+            app_memory: OptionalCell::empty(),
+            flash_start: Cell::new(0),
+            flash_end: Cell::new(0),
+            fault_policy,
+            new_process_flash: OptionalCell::empty(),
+            driver: driver,
+            buffer: TakeCell::new(buffer),
+            new_app_start_addr: Cell::new(0),
+            new_app_length: Cell::new(0),
+            client: OptionalCell::empty(),
+            current_user: OptionalCell::empty(),
+        }
+    }
+
+    // Needs to be set separately, or breaks grants allocation
+    pub fn flash_and_memory(
+        &self,
+        app_memory: &'static mut [u8],
+        flash_start: usize,
+        flash_end: usize,
+    ) {
+        self.app_memory.set(app_memory);
+        self.flash_start.set(flash_start);
+        self.flash_end.set(flash_end);
+    }
+
+    // Function to find the next available slot in the processes array
+    fn find_open_process_slot(&self) -> Option<usize> {
+        self.procs.map_or(None, |procs| {
+            for (i, p) in procs.iter().enumerate() {
+                if p.is_none() {
+                    return Some(i);
+                }
+            }
+            None
+        })
+    }
+
+    /******************************************* NVM Stuff **********************************************************/
+
+    // Check so see if we are doing something. If not, go ahead and do this
+    // command. If so, this is queued and will be run when the pending
+    // command completes.
+    fn enqueue_command(
```

Comment:
I don't see where this enqueues anything. I think either we can do the read/write or this is an error.

Reply:
Yeah, changed it.